ch-covidcertificate-backend-verifier/ch-covidcertificate-backend-verifier-data/src/main/java/ch/admin/bag/covidcertificate/backend/verifier/data/impl/JdbcVerifierDataServiceImpl.java

```diff
@@ -45,41 +52,104 @@ public JdbcVerifierDataServiceImpl(DataSource dataSource) {
     }
 
     @Override
+    @Transactional
     public void insertCscas(List<DbCsca> cscas) {
-        // TODO
+        logger.debug(
+                "Inserting CSCA certificates with kid's: {}",
+                cscas.stream().map(DbCsca::getKeyId).collect(Collectors.toList()));
+        List<SqlParameterSource> batchParams = new ArrayList<>();
+        if (!cscas.isEmpty()) {
+            for (DbCsca dbCsca : cscas) {
+                batchParams.add(getCSCAParams(dbCsca));
+            }
+            cscaInsert.executeBatch(
+                    batchParams.toArray(new SqlParameterSource[batchParams.size()]));
+        }
+    }
+
+    @Override
+    @Transactional
+    public int removeCSCAs(List<String> keyIds) {
+        if (!keyIds.isEmpty()) {
+            var sql = "delete from t_country_specific_certificate_authority";
+            final var params = new MapSqlParameterSource();
+            sql += " where key_id in (:kids)";
+            params.addValue("kids", keyIds);
+            return jt.update(sql, params);
+        } else {
+            return 0;
+        }
+    }
+
+    @Override
+    @Transactional(readOnly = true)
+    public List<DbCsca> findCscas(String origin) {
+        final var sql =
+                "select * from t_country_specific_certificate_authority where origin = :origin";
+        final var params = new MapSqlParameterSource();
+        params.addValue("origin", origin);
+        return jt.query(sql, params, new CSCARowMapper());
     }
 
     @Override
-    public int removeCscasNotIn(List<String> keyIdsToKeep) {
-        // TODO
-        return 0;
+    @Transactional(readOnly = true)
+    public List<String> findActiveCscaKeyIds() {
+        final var sql = "select key_id from t_country_specific_certificate_authority";
+        return jt.queryForList(sql, new MapSqlParameterSource(), String.class);
     }
 
     @Override
-    public void insertDsc(DbDsc dsc) {
-        // TODO
+    @Transactional
+    public void insertDsc(List<DbDsc> dsc) {
```

Comment:
```suggestion
    public void insertDscs(List<DbDsc> dsc) {
```
every function before was with plural "s", but here it is missing